security/advancedtls/crl_provider_test.go

```diff
@@ -0,0 +1,181 @@
+		crl := loadCRL(t, testdata.Path(fmt.Sprintf("crl/%d.crl", i)))
+		p.AddCRL(crl)
+	}
+	// Each test data entry contains a description of a certificate chain,
+	// certificate chain itself, and if CRL is not expected to be found.
+	tests := []struct {
+		desc        string
+		certs       []*x509.Certificate
+		expectNoCRL bool
+	}{
+		{
+			desc:  "Unrevoked chain",
+			certs: makeChain(t, testdata.Path("crl/unrevoked.pem")),
+		},
+		{
+			desc:  "Revoked Intermediate chain",
+			certs: makeChain(t, testdata.Path("crl/revokedInt.pem")),
+		},
+		{
+			desc:  "Revoked leaf chain",
+			certs: makeChain(t, testdata.Path("crl/revokedLeaf.pem")),
+		},
+		{
+			desc:        "Chain with no CRL for issuer",
+			certs:       makeChain(t, testdata.Path("client_cert_1.pem")),
+			expectNoCRL: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			for _, c := range tt.certs {
+				crl, err := p.CRL(c)
+				if err != nil {
+					t.Fatalf("Expected error fetch from provider: %v", err)
+				}
+				if crl == nil && !tt.expectNoCRL {
+					t.Fatalf("CRL is unexpectedly nil")
+				}
+			}
+		})
+	}
+}
+
+// TestFileWatcherCRLProviderConfig checks creation of FileWatcherCRLProvider,
+// and the validation of Options configuration. The configurations include empty
+// one, non existing CRLDirectory, invalid RefreshDuration, and the correct one.
+func TestFileWatcherCRLProviderConfig(t *testing.T) {
+	if _, err := MakeFileWatcherCRLProvider(Options{}); err == nil {
+		t.Fatalf("Empty Options should not be allowed")
+	}
+	if _, err := MakeFileWatcherCRLProvider(Options{CRLDirectory: "I_do_not_exist"}); err == nil {
+		t.Fatalf("CRLDirectory must exist")
+	}
+	if defaultProvider, err := MakeFileWatcherCRLProvider(Options{CRLDirectory: testdata.Path("crl/provider")}); err == nil {
+		if defaultProvider.opts.RefreshDuration != defaultCRLRefreshDuration {
+			t.Fatalf("RefreshDuration is not properly updated by validate() func")
+		}
+		defaultProvider.Close()
+	} else {
+		t.Fatal("Unexpected error:", err)
+	}
+
+	customCallback := func(err error) {
+		fmt.Printf("Custom error message: %v", err)
+	}
+	regularProvider, err := MakeFileWatcherCRLProvider(Options{
+		CRLDirectory:               testdata.Path("crl"),
+		RefreshDuration:            5 * time.Second,
+		cRLReloadingFailedCallback: customCallback,
+	})
+	if err != nil {
+		t.Fatal("Unexpected error while creating regular FileWatcherCRLProvider:", err)
+	}
+	regularProvider.Close()
+}
+
+// TestFileWatcherCRLProvider tests how FileWatcherCRLProvider handles the major
+// four cases for CRL checks. It scans the CRLs under crl directory to populate
+// the in-memory storage. Then we construct unrevoked, revoked leaf, and revoked
+// intermediate chains, as well as a chain without CRL for issuer, and check
+// that it’s correctly processed. Additionally, we also check if number of
+// invocations of custom callback is correct.
+func TestFileWatcherCRLProvider(t *testing.T) {
```

Comment:
Do we have a test verifying the async refreshing goroutine works?

Reply:
Not yet, it'll require playing with the timers so I'm planning to have a separate PR for that (like e2e test)